backend/internal/database/book.go

```diff
@@ -297,9 +297,129 @@ func (db *BookDatabase) IncrementBookCounter(ctx context.Context, id string) err
+	}
+
+	err = rows.Scan(&analytics)
+
+	if err != nil {
+		return analytics, errors.Wrap(err, "error on Scan for book clicks array")
+	}
+
+	return analytics, err
+
+}
+
+/*
+ * Returns a dictionary mapping each book's id to its daily click counts given a range from the current day
+ */
+func (db *BookDatabase) FetchAllBookAnalytics(ctx context.Context, dRange int) ([]models.Analytic, error) {
+	analytics := make([]models.Analytic, 0)
+
+	var firstYearStart int
+	var firstYearEnd int
+	var secondYearEnd int
+
+	firstYearStart, firstYearEnd, secondYearEnd = AnalyticsClicksIndices(dRange)
+
+	var query string = "SELECT id, ARRAY_CAT(clicks[$1:$2], clicks[1:$3]) " +
+		"FROM book_analytics"
+	rows, err := db.Conn.Query(ctx, query, firstYearStart, firstYearEnd, secondYearEnd)
+
+	if err != nil {
+		fmt.Print(err)
+		return analytics, errors.Wrap(err, "error on query for all book analytics")
+	}
+
+	defer rows.Close()
+
+	for rows.Next() {
+		var analytic models.Analytic
+		if err := rows.Scan(&analytic.ID, &analytic.Clicks); err != nil {
+			fmt.Print(err)
+			return analytics, errors.Wrap(err, "error on Scan for book id and clicks array in "+
+				"FetchAllBookAnalytics")
+		}
+
+		analytics = append(analytics, analytic)
+	}
+
+	return analytics, err
+
+}
+
+/*
+ * Gets 5 most popular books from the past 30 days based on click analytics
+ */
+func (db *BookDatabase) FetchPopularBooks(ctx context.Context) ([]models.Book, error) {
+	books := make([]models.Book, 0)
+
+	var dRange int = 30
+	var firstYearStart int
+	var firstYearEnd int
+	var secondYearEnd int
+
+	firstYearStart, firstYearEnd, secondYearEnd = AnalyticsClicksIndices(dRange)
+	var query string = "SELECT books.id, title, author, image, created_at, " +
+		"array_remove(array_agg(lang), NULL) as languages " +
+		"FROM books LEFT JOIN book_contents ON books.id = " +
+		"book_contents.id LEFT JOIN book_analytics ON books.id = book_analytics.id " +
+		"GROUP BY books.id, book_analytics.id " +
+		"ORDER BY (SELECT SUM(analytics) " +
+		"FROM UNNEST(ARRAY_CAT(clicks[$1:$2], clicks[1:$3])) AS analytics) " +
+		"DESC FETCH FIRST 5 ROWS ONLY"
+	rows, err := db.Conn.Query(ctx, query, firstYearStart, firstYearEnd, secondYearEnd)
+
+	if err != nil {
+		fmt.Print(err)
+		return books, errors.Wrap(err, "error on query for most popular books")
+	}
+
+	defer rows.Close()
+
+	for rows.Next() {
+		var book models.Book
+		if err := rows.Scan(&book.ID, &book.Title, &book.Author, &book.Image,
+			&book.CreatedAt, &book.Languages); err != nil {
+			fmt.Print(err)
+			return books, errors.Wrap(err, "error on Scan for book information in "+
+				"FetchPopularBooks")
+		}
+
+		books = append(books, book)
+	}
+
+	return books, err
+}
+
+/*
+ * Helper function that returns indices used to slice into clicks array
+ * to select the past <dRange> days of analytics
+ *
+ * ex: ARRAY_CAT(clicks[firstYearStart:firstYearEnd], clicks[1:secondYearEnd])
+ */
+func AnalyticsClicksIndices(dRange int) (int, int, int) {
```

Comment:
I think this logic can be moved into SQL pretty easily with the expression I DMed you both.

`clicks[366-30+TODAY:] || clicks[TODAY-30:TODAY]` will always select the last 30 days with wraparound (as far as I can tell)